staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/types.go

```diff
@@ -104,24 +104,19 @@ func (q *queue) Dequeue() (*request, bool) {
 // GetNextFinish returns the expected virtual finish time of the
 // oldest request in the queue with estimated duration G
 func (q *queue) GetNextFinish(G float64) float64 {
-	// TODO: if we decide to generalize this function to return virtual finish time
-	//  for the Nth oldest request waiting in the queue, we need to carefully
-	//  evaluate and potentially improve the performance here.
-	var oldestReq *request
-	q.requests.Walk(func(r *request) bool {
-		oldestReq = r
-		return false
-	})
-	if oldestReq == nil {
+	if q.requests.Length() == 0 {
 		// we should never be here, since the caller should ensure
 		// that this queue has request(s) waiting to be served before
 		// calling this function.
 		return q.virtualStart
 	}
 
-	// the estimated service time of the oldest request is (G * request width)
-	estimatedServiceTime := float64(G) * float64(oldestReq.Seats())
-	return q.virtualStart + estimatedServiceTime
+	// the estimated service time of the oldest request is:
+	//   virtual start time + G
+	// We are not taking the width of the request into account when
+	// we calculate the virtual finish time of the request because
+	// it can starve requests with smaller wdith in other queues.
+	return q.virtualStart + float64(G)
```

Comment:
I think in the current form we have that inconsistent with other places where we actually modify virtualStart with "G x Seats()".

I would actually suggest changing this method to "GetNextStart() float64" and simply returning q.virtualStart.

Given that we only use this function in one place and only comparison matters (and we use common G everywhere), I it would actually be much more intuitive and consistent with all other places.

Reply:
actually i just removed it for now, since it's a one-liner.